File: server-catalog/src/Catalog.Api/Dtos/ErrorEnvelopeDto.cs
using System.Text.Json.Serialization;

namespace Catalog.Api.Dtos;

/// <summary>
/// The one error shape this API ever returns -- ExceptionHandlingMiddleware (unhandled
/// exceptions) and the rate limiter's OnRejected handler (429s, see RateLimitingExtensions)
/// both serialize this same type, so there's exactly one envelope to document in Swagger and
/// exactly one to handle on the client, not one ad hoc shape per failure path. Mirrors
/// server-orders' AllExceptionsFilter envelope field-for-field (see architecture.md's
/// cross-service error contract) -- [JsonPropertyName] keeps the wire format lowercase
/// (status/message/errors) despite the PascalCase C# property names.
/// </summary>
public sealed class ErrorEnvelopeDto
{
    [JsonPropertyName("status")]
    public int Status { get; set; }

    [JsonPropertyName("message")]
    public string Message { get; set; } = string.Empty;

    [JsonPropertyName("errors")]
    public string[] Errors { get; set; } = Array.Empty<string>();
}

File: server-orders/src/common/dto/error-envelope.dto.ts
import { ApiProperty } from '@nestjs/swagger';

/**
 * The one error shape this API ever returns -- AllExceptionsFilter serializes exactly this
 * for every error response, whatever raised it (validation, an unhandled exception, the
 * throttler, ConcurrencyLimitMiddleware's manual ServiceUnavailableException). A class with
 * @ApiProperty decorators (not the bare `ErrorEnvelope` interface AllExceptionsFilter itself
 * uses) so @nestjs/swagger can actually generate a schema from it for @ApiResponse -- Swagger
 * can't reflect field names/types out of a TypeScript-only interface, which has no runtime
 * representation once compiled away. Mirrors server-catalog's ErrorEnvelopeDto field-for-field
 * (see architecture.md's cross-service error contract).
 */
export class ErrorEnvelopeDto {
  @ApiProperty()
  status: number;

  @ApiProperty()
  message: string;

  @ApiProperty({ type: [String] })
  errors: string[];
}
